### PR TITLE
En 10295 retry ports on net messenger

### PR DESCRIPTION
**Pull request by iulianpascalau**
- implemented port retrial mechanism that will prevent node creation errors in extremely rare situations

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -32,6 +32,8 @@ require (
 	github.com/libp2p/go-libp2p-kad-dht v0.12.2
 	github.com/libp2p/go-libp2p-kbucket v0.4.7
 	github.com/libp2p/go-libp2p-pubsub v0.4.1
+	github.com/libp2p/go-libp2p-transport-upgrader v0.4.2
```
Comment (bogdan-rosianu) [inline go.mod:35]: `go mod tidy` and maybe commit go sum also?

File: node/mock/messengerStub.go [hunk 2]
```diff
@@ -184,3 +185,12 @@ func (ms *MessengerStub) GetConnectedPeersInfo() *p2p.ConnectedPeersInfo {
 func (ms *MessengerStub) UnjoinAllTopics() error {
 	return nil
 }
+
+// Port -
+func (ms *MessengerStub) Port() int {
```
Comment (bogdan-rosianu) [inline node/mock/messengerStub.go:190]: could move the stub to testscommon

File: p2p/libp2p/netMessenger.go [hunk 3]
```diff
@@ -130,39 +139,78 @@ func NewNetworkMessenger(args ArgsNetworkMessenger) (*networkMessenger, error) {
...
+			return h, ctx, cancelFunc, port, nil
+		}
+
+		lastErr = err
+		if !strings.Contains(err.Error(), msgBindError) {
+			//not a bind error, return directly
+			return nil, nil, nil, 0, err
+		}
+
+		log.Debug("bind error in network messenger", "port", port, "retry number", i)
```
Comment (SebastianMarian) [inline p2p/libp2p/netMessenger.go:210]: Could be i + 1

File: p2p/libp2p/netMessenger.go [hunk 4]
```diff
@@ -115,6 +120,10 @@ type ArgsNetworkMessenger struct {
 
 // NewNetworkMessenger creates a libP2P messenger by opening a port on the current machine
 func NewNetworkMessenger(args ArgsNetworkMessenger) (*networkMessenger, error) {
+	return newNetworkMessenger(args, true, true)
+}
+
+func newNetworkMessenger(args ArgsNetworkMessenger, withMessageSigning bool, reusePort bool) (*networkMessenger, error) {
```
Comment (bogdan-rosianu) [inline p2p/libp2p/netMessenger.go:126]: can you replace the bool parameters with types? 
```
type messageSigningConfig bool
const (
    withMessageSigning messageSigningConfig = true
    withoutMessageSigning messageSigningConfig = true
)
```
or something similar

**Review by SebastianMarian — APPROVED** [review]

File: p2p/libp2p/netMessenger.go [hunk 5]
```diff
@@ -130,39 +139,78 @@ func NewNetworkMessenger(args ArgsNetworkMessenger) (*networkMessenger, error) {
...
+	p2pNode, err := createMessenger(args, h, ctx, cancelFunc, port, withMessageSigning)
+	if err != nil {
+		log.LogIfError(h.Close())
+		return nil, err
+	}
+
+	return p2pNode, nil
+}
+
+func createHost(args ArgsNetworkMessenger, p2pPrivKey *libp2pCrypto.Secp256k1PrivateKey, portReuse bool) (host.Host, context.Context, func(), int, error) {
```
Comment (bogdan-rosianu) [inline p2p/libp2p/netMessenger.go:158]: too many returned stuff. maybe use a DTO for returning and you'd use something like `(createHostResults, error)`

File: p2p/libp2p/netMessenger_test.go [hunk 6]
```diff
@@ -1693,6 +1693,48 @@ func TestNetworkMessenger_ApplyOptionsShouldWork(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestNetworkMessenger_ChooseAnotherPortIfBindFails(t *testing.T) {
+	port := "37000-37100" //use a larger interval as some ports might already be used by the OS
```
Comment (bogdan-rosianu) [inline p2p/libp2p/netMessenger_test.go:1697]: skip this test by default? it might fail or take long times on some machines

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: p2p/libp2p/netMessenger.go (hunk 3 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go:210]: changed

File: go.mod (hunk 1 above)
Comment (iulianpascalau) [reply go.mod:35]: done, nothing changed.

File: node/mock/messengerStub.go (hunk 2 above)
Comment (iulianpascalau) [reply node/mock/messengerStub.go:190]: done, removed duplicated stubs, improved the only common messenger stub

File: p2p/libp2p/netMessenger_test.go (hunk 6 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1697]: right, will be disabled by default

File: p2p/libp2p/netMessenger.go (hunk 4 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go:126]: done

File: p2p/libp2p/netMessenger_test.go (hunk 6 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger_test.go:1697]: done

File: p2p/libp2p/netMessenger.go (hunk 5 above)
Comment (iulianpascalau) [reply p2p/libp2p/netMessenger.go:158]: refactored

**Review by SebastianMarian — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.

**Pull request merged**